planner/cascades/transformation_rules.go

```diff
@@ -32,12 +39,135 @@ func GetTransformationRules(node plannercore.LogicalPlan) []Transformation {
+}
+
+// Match implements Transformation interface.
+func (r *PushSelDownTableScan) Match(expr *memo.ExprIter) bool {
+	return true
+}
+
+// OnTransform implements Transformation interface.
+func (r *PushSelDownTableScan) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
+	ts := old.Children[0].GetExpr().ExprNode.(*plannercore.TableScan)
+	if ts.Handle == nil {
+		return nil, false, false, nil
+	}
+	accesses, remained := ranger.DetachCondsForColumn(ts.SCtx(), sel.Conditions, ts.Handle)
+	if accesses == nil {
+		return nil, false, false, nil
+	}
+	newTblScan := plannercore.TableScan{
+		Source:      ts.Source,
+		Handle:      ts.Handle,
+		AccessConds: ts.AccessConds.Shallow(),
+	}.Init(ts.SCtx())
+	newTblScan.AccessConds = append(newTblScan.AccessConds, accesses...)
+	tblScanExpr := memo.NewGroupExpr(newTblScan)
+	expr := tblScanExpr
+	if len(remained) > 0 {
+		schema := old.GetExpr().Group.Prop.Schema
+		tblScanGroup := memo.NewGroupWithSchema(tblScanExpr, schema)
+		newSel := plannercore.LogicalSelection{
+			Conditions: remained,
+		}.Init(sel.SCtx())
+		selExpr := memo.NewGroupExpr(newSel)
+		selExpr.Children = append(selExpr.Children, tblScanGroup)
+		expr = selExpr
+	}
+	newExprs = append(newExprs, expr)
+	return newExprs, true, false, nil
+}
+
+// PushSelDownTableGather pushes the selection down to child of TableGather.
+type PushSelDownTableGather struct {
+}
+
+// GetPattern implements Transformation interface. The pattern of this rule is `Selection -> TableGather`
+func (r *PushSelDownTableGather) GetPattern() *memo.Pattern {
+	any := memo.NewPattern(memo.OperandAny)
+	tg := memo.BuildPattern(memo.OperandTableGather, any)
+	return memo.BuildPattern(memo.OperandSelection, tg)
+}
+
+// Match implements Transformation interface.
+func (r *PushSelDownTableGather) Match(expr *memo.ExprIter) bool {
+	return true
+}
+
+// OnTransform implements Transformation interface.
+func (r *PushSelDownTableGather) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
+	tg := old.Children[0].GetExpr().ExprNode.(*plannercore.TableGather)
+	childGroup := old.Children[0].Children[0].Group
+	var pushed, remained []expression.Expression
+	ctx := tg.SCtx()
+	_, pushed, remained = expression.ExpressionsToPB(ctx.GetSessionVars().StmtCtx, sel.Conditions, ctx.GetClient())
+	if len(pushed) == 0 {
+		return nil, false, false, nil
+	}
+	pushedSel := plannercore.LogicalSelection{Conditions: pushed}.Init(ctx)
+	pushedSelExpr := memo.NewGroupExpr(pushedSel)
+	pushedSelExpr.Children = append(pushedSelExpr.Children, childGroup)
+	pushedSelGroup := memo.NewGroupWithSchema(pushedSelExpr, childGroup.Prop.Schema)
+	// Currently field content of TableGather would not be modified, so we just reference same tg instead of making
```

Comment:
how about:
```go
    // The field content of TableGather would not be modified currently, so we
    // just reference the same tg instead of making a copy of it.
    //
    // TODO: if we save the pushed filters later in TableGather, in order to do
    //       partition pruning or skyline pruning, we need to make a copy of
    //       the TableGather here.
```